Code_Cleanup_NewPR

birdday opened the pull request:
This is all the same changes as before, except with the added final change of reverting all comments back to a doc string format for use as a help message. A lot of rebasing has been done to make the changes easier to track. There were still a few parts that were slightly confusing, but I think it should be much better all around. We can probably workshop on this version for any necessary changes, but I found a new PR easier than changing the last one I submitted.

birdday committed:
Changes to the yaml and sample yaml files to include all 50 MOFS.

Also includes some changes to the .gitignore file to temporarily track and then untrack these files for later iterations.

birdday committed:
Reorganized functions, no change to actual code

birdday committed:
Added an alternate way of saving unbinned pmf data.

This was done for easier plotting later on. The original method is still preserved in case its needed for benchmarking to any of Jenna's old results. Can be deleted in the future if desired.

birdday committed:
Changed all function comments into docstring type comments

birdday committed:
Changes to the calculation of pmf values for a single array, calculate_array_pmf.

Function is not called externally, only in calculate_all_arrays, hence no changes to execute script. Name change is other function is part of the next commit (left separate to reduce content of a single commit).

birdday committed:
Changes to calculation of arrays and array pmfs, calculate_all_arrays.

Name change reflected in execute script.

birdday committed:
Began to modify interpolation function. Currently not functional or called, but is intended to be revisted for later use.

birdday committed:
Changes to bin creation, create_bins.

Changed the creation of bins so that the boundaries fall in the middle of simulated compositions.
Left Jenna's bin calculations in in case they are needed for comparison in the short term. Can/will delete later.
Changes to create_bins reflected in execute script.

birdday committed:
Changes to binning procedure, bin_compositions.

Minor changes to binning procedure for clarity.
Changes to bin_compositons reflected in execute script.

birdday committed:
-- plot_binned_pmf_array --

birdday committed:
-- save_array_pmf_data --

Mild changes to the format of saved data for ease of use later. Also changed names of folders for saved data to imporve clarity.

birdday committed:
-- save_element_pmf_data --

Changed the format of the pmf data for each MOF for ease of use later. Function isn't needed beyond a single run, regardless of MOF size, but is useful.

birdday committed:
Changed the calculation of KLD scores, calculate_kld.

Introduced and error in which kld scores were handled as strings rather than floats, so sorting and selection did not work as expected. That fix is now squashed here. Leaving message in commit just in case since debugging this took a while the first time.
Reflected changes to function in execute script.

birdday committed:
Changes to ranking and saving of arrays, choose_arrays.

Reflected changes to function in execute script.

birdday committed:
Changes to handling of yaml file in execute script.

birdday committed:
Changes to saving /plotting in execute script.

Added section header for saving / plotting.
Renamed functions for saving/plotting array_pmf data and reorganized execute process mass data code marginally

birdday committed:
Choose arrays based on absolute KLD added to execute script.

birdday committed:
Changed to pmf calculation for a single mof, calculate_element_pmf.

birdday committed:
Changed the description at the top of the code, process_mass_data.py

Brief overview of Jenna's work/goals and my own. Added a list of all keywords.

birdday committed:
Regorganize imported modules alphabetically, and by import vs. from.

birdday committed:
Changes to function names which read/write data.

Added a small section header for user defined functions.
Note that the change from sys.argv to variables is reflected in a later commit. For some reason, could not reorder or squash that particular commit. Variable now finds data files in its path, but can be easily reverted to a sys.argv if desired.

birdday committed:
Changed the functions which import the experimental and simulated data.

Should eliminate one instance of these functions later and rewrite so that a single version could be used for importing both exp and sim data.
No changes to yaml laod function, though I think github tracks it as being deleted and re-added.
Changed names of import functions in execute script.

birdday committed:
Changed description at top of code, execute_process_mass_data.py

birdday committed:
Changes to paths for data files. No longer sys.argv.

birdday committed:
Changes to how / where data is saved. Included array ID generation.

birdday committed:
Changes to calculation of KLD scores, calculate_kld.

birdday committed:
Changes to ranking and saving of arrays, choose_arrays.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -56,14 +46,28 @@
+from itertools import combinations
+from math import isnan, log
+from matplotlib import pyplot as plt
+from random import random
```
PEP8 (https://www.python.org/dev/peps/pep-0008/#imports says to group them by:

Standard library imports.
Related third party imports.
Local application/library specific imports.
You should put a blank line between each group of imports.

This is so you can tell which libraries should always succeed imports, which that require a pip install, and which are specific to this module.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -220,23 +220,23 @@ def calculate_element_pmf(exp_results_full, sim_results_full, mof_list, stdev, m
+# Function is used in function 'calculate_all_arrays' below
+# Keyword arguments:
+#     mof_array -- list of mofs in a single array
+#     element_pmf_results -- list of dictionaries including mof, mixture, probability
```
In general, if you were going to make all the comment changes in a final commit, you wouldn't change them here too. But OK.

paulboone replied:
95f62515ccbc595298113ceacc4e9f190d7131ed

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -220,23 +220,23 @@ def calculate_element_pmf(exp_results_full, sim_results_full, mof_list, stdev, m
-    mof_array -- list of mofs in array
-    calculate_pmf_results -- list of dictionaries including mof, mixture, probability
-    """
+# ----- Combines and normalizes mof pmfs for a single array -----
```
"Name change is other function is part of the next commit (left separate to reduce content of a single commit)."

Using separate commits simply in order to reduce content of single commit is not a valid justification. Plus I would expect that making the change in this commit would be an addition of +1 line.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -220,23 +220,23 @@ def calculate_element_pmf(exp_results_full, sim_results_full, mof_list, stdev, m
             compound_pmfs = mof_pmf
+    norm_factor = sum(compound_pmfs)
+    single_array_pmf_results = [ i / norm_factor for i in compound_pmfs ]
+    return(single_array_pmf_results)
```
code after return should be deleted.

paulboone replied:
Why are there parentheses around single_array_pmf_results? It is weird to return a tuple of one item.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -220,23 +220,23 @@ def calculate_element_pmf(exp_results_full, sim_results_full, mof_list, stdev, m
-        mof_pmf = [ row['PMF'] for row in calculate_pmf_results if row['MOF'] == mof ]
-
-        # Joint prob, multiply pmf values elementwise
+        mof_pmf = [ row['PMF_Range'] for row in element_pmf_results if row['MOF'] == mof ]
```
row['PMF_Range'] was formerly row['PMF'] which means this code is either a bug fix (and the code never worked), or you've broken the code with this commit because you changed how it was accessed without changing how it was set.

birdday replied:
I changed how it was set in a different commit, but did not catch that when rebasing, so you're right that this is broken code as of this commit - my bad. Can fix if needed.

paulboone commented on execute_process_mass_data.py:
```diff
@@ -84,6 +83,8 @@
 (element_pmf_results) = (
     calculate_element_pmf(exp_results_full, sim_results_full, mof_list, stdev, mrange) )
+(list_of_arrays, all_array_pmf_results) = (
+    calculate_all_arrays(mof_list, num_mofs, element_pmf_results, gases) )
```
Delete extraneous parentheses. Please delete the ones above too that aren't part of this commit, since they are super weird.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -238,55 +238,43 @@ def calculate_array_pmf(mof_array, element_pmf_results):
-                array_pmf.extend([array_dict])
+            # First Array: Set up dictionary with keys
+            for index in range(len(comp_set_dict)):
+                array_dict = {'%s' % ' '.join(mof_array) : single_array_pmf_results[index]}
```
'%s' % ' '.join(mof_array) => ' '.join(mof_array)

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -238,55 +238,43 @@ def calculate_array_pmf(mof_array, element_pmf_results):
+            for index in range(len(comp_set_dict)):
+                array_dict = {'%s' % ' '.join(mof_array) : single_array_pmf_results[index]}
+                for gas in gases:
+                    array_dict.update({ '%s' % gas : float(comp_set_dict[index][gas])})
```
if gas is already a string, just use gas. If gas isn't a string, use str(gas)

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -238,55 +238,43 @@ def calculate_array_pmf(mof_array, element_pmf_results):
-                array_pmf[index]['%s' % ' '.join(mof_array)] = normalized_compound_pmfs[index]
+            # Not First Array: Update Dictionary
+            for index in range(len(comp_set_dict)):
+                all_array_pmf_results[index]['%s' % ' '.join(mof_array)] = single_array_pmf_results[index]
```
'%s' % ' '.join(mof_array) => ' '.join(mof_array)

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -238,55 +238,43 @@ def calculate_array_pmf(mof_array, element_pmf_results):
+                all_array_pmf_results[index]['%s' % ' '.join(mof_array)] = single_array_pmf_results[index]
 
-    return(array_pmf, mof_array_list)
+    return(mof_array_list, all_array_pmf_results)
```
add space between return and (

paulboone commented on execute_process_mass_data.py:
```diff
@@ -85,6 +84,8 @@
 (list_of_arrays, all_array_pmf_results) = (
     calculate_all_arrays(mof_list, num_mofs, element_pmf_results, gases) )
+(bins) = (
+    create_bins(gases, num_bins, mof_list, element_pmf_results) )
```
remove extraneous parentheses

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -276,24 +276,45 @@ def calculate_all_arrays(mof_list, num_mofs, element_pmf_results, gases):
+        # lower_lim = lower_comp
+        # upper_lim = upper_comp + (upper_comp-lower_comp)/(num_bins)
+        bin_points_temp = np.linspace(lower_lim, upper_lim, num=num_bins+1, endpoint=True)
+        # Extra steps to deal with impercision of values generated by np.linspace
```
sp: imprecision.

Can you elaborate on what the imprecision is with np.linspace and why the next few lines are supposed to fix it?

birdday replied:
I had noticed that about 9-12 digits out I would see non-zero digits (i.e. 0.3000000000007 rather than 0.30). Although this should be fine when my bin spacing is halfway between the relevant compositions, which they currently are, when a bin and a bin point fall on the exact same value, which they might if I reintroduce interpolation or have some reason to use larger or smaller bins, the mix between bin values like 0.300000007 and 0.30 would make using if x >= y or (<= or ==) shift certain points into bins where I did not want them. I can see if there is a way to specify the precision of np.linspace, or just adjust the precision of an array, but this was just the fix I came up with first.

paulboone replied:
Are you familiar with how floating point numbers are stored? You shouldn't expect to have even boundaries; they will always be a very, very tiny fraction off from the "correct" number. It shouldn't rarely, rarely, rarely (if ever) matter though. If you adjust the precision, you will just have the errors be more digits out, since floating point numbers are not exact representations. But it really shouldn't matter though. Why does it matter if a 0.30...0x value goes into the <0.3 bin or the >= 0.3 bin? Isn't it all the same?

paulboone replied:
Ok, so take a look at linspace, and if you can replicate the error, we can take a look about fixing it there, and then remove the bin shift and the bin rounding.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -276,24 +276,45 @@ def calculate_all_arrays(mof_list, num_mofs, element_pmf_results, gases):
+    # Reformat bin_points
+    bins = []
+    for row in bin_points:
+        bins.append({gases[i] : row[i] for i in range(len(gases))})
```
This is fine, but I'm a fan of the list comprehensions when they fit in one line. They are also faster.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -318,55 +318,64 @@ def create_bins(gases, num_bins, mof_list, element_pmf_results):
+            for array in array_names:
+                if array_pmfs_temp[array] == []:
+                    with_sum.update({array : 0})
+                    with_max.update({array : 0})
```
Can we just use normal syntax: `with_sum[array] = 0` ?

birdday replied:
The temporary dictionary which is being updated, single_bin_temp, does not yet include the key for the array, so there would be a key error, hence the need for update. 

I know we talked about pre-allocating all these keys, or switching to an object for managing this data, but I was going to save that for a separate exercise / set of commits on a new feature branch following this PR.

paulboone replied:
`dictionary[key] = value` doesn't require that the key exist in the dictionary before you use it. Typically you would only use update when you wanted to merge two dictionaries, or if you had a lot of key-value pairs you wanted to update in your dictionary, but if you only have one key-value pair to set, you typically use the standard `dictionary[key] = value` syntax. 

If you are going to rewrite this, I'm ok with you leaving it, as long as my above point is clear...?

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -318,55 +318,64 @@ def create_bins(gases, num_bins, mof_list, element_pmf_results):
+        binned_probabilities_max.extend(all_bins_temp_max)
 
-    return(binned_probability)
+    return(binned_probabilities_sum, binned_probabilities_max)
```
return is a statement, not a method, so use a space after it

paulboone reviewed (changes requested):
I only got through 8 commits – PHEW. Some of it we need to review together for sure. If you are in tomorrow, we can do it then, or we can use the time-tested technique of doing code reviews over the phone. In general, the commits are a LOT better this time!

birdday committed:
Removed all parenthesis associated with 'return' in functions. Nothing needed to be a tuple.

birdday committed:
Removed extra parenthesis during function calls, but added line continuation in the form of '/'.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -343,31 +343,60 @@ def bin_compositions(gases, list_of_arrays, create_bins_results, array_pmf_resul
+        array_id = str(num_elements)+'-'+str(i)
+        array_name = '%s' % ' '.join(array)
+        array_id_key.append([array_id, array_name])
+        array_id_dict.update({array_name: array_id})
```
This should be a separate function; line 364-369 should be rethought or at minimum commented; `num_elements = num_elements` should be deleted;

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -399,25 +399,49 @@ def plot_binned_pmf_array(gases, list_of_arrays, bins, binned_probabilities):
+    with open(filename,'w', newline='') as csvfile:
+        writer = csv.writer(csvfile, delimiter="\t")
+        for key, val in array_id_dict.items():
+            writer.writerow([key, val])
```
like before, should be made into a separate method

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -448,18 +448,13 @@ def save_array_pmf_data(gases, list_of_arrays, bins, binned_probabilities):
+def save_element_pmf_data(element_pmf_results, stdev, mrange):
+    timestamp = datetime.now().strftime("%Y_%m_%d__%H_%M_%S")
+    data_frame = pd.DataFrame(element_pmf_results)
+    data_frame.to_csv('saved_element_pmfs/%s_stdev_%s_mrange_%s.csv' % (stdev, mrange, timestamp), sep='\t')
```
has to have calling method in same commit

paulboone commented on execute_process_mass_data.py:
```diff
@@ -55,23 +55,23 @@
-    mof_densities_import.copy()
-    mof_densities_import.update({ mof : data['mofs'][mof]['density']})
-
+# Redefine key varaibles in yaml file
```
sp: variables

paulboone commented on execute_process_mass_data.py:
```diff
@@ -55,23 +55,23 @@
+num_mofs = data['number_mofs']
 num_mixtures = data['num_mixtures']
+num_bins = data['num_bins']
+num_best_worst = data['num_best_worst']
```
not added anywhere? should be in this commit.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -147,62 +147,78 @@ def add_random_gas(gases, comps, num_mixtures):
+            sum_probs_range = sum(probs_range)
+            norm_probs_range = [(i/sum_probs_range) for i in probs_range]
+            # sum_probs_exact = sum(probs_exact)
+            # norm_probs_exact = [(i/sum_probs_exact) for i in probs_exact]
```
should usually never have commented out code, but if you do, you must write an explanatory comment explaining what it is for why

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -79,92 +79,55 @@ def read_data_as_dict(filename):
+#     gases -- list of gases in simulated mixtures
+def import_experimental_data(exp_results_import, mof_list, mof_densities, gases):
+    # Can probibily omit exp_results_mass in the future and just use the full
+    # dictionary for furthere analysis.
```
probibily; furthere

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -79,92 +79,55 @@ def read_data_as_dict(filename):
-            simulated_results.extend([temp_dict])
+                mass = float(mof_densities[mof]) * float(row['Mass'])
+                exp_results_temp = row.copy()
+                exp_results_temp.update({'Mass_mg/cm3' : mass})
```
use other [] syntax instead of update

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -79,92 +79,55 @@ def read_data_as_dict(filename):
+            if row['MOF'] == mof:
+                mass = float(mof_densities[mof]) * float(row['Mass'])
+                sim_results_temp = row.copy()
+                sim_results_temp.update({'Mass_mg/cm3' : mass})
```
update

paulboone commented on execute_process_mass_data.py:
```diff
@@ -1,21 +1,47 @@
-                                            save_raw_pmf_data,
-                                            information_gain,
-                                            choose_best_arrays)
+from mof_array.pmf.process_mass_data import *
```
never use import *

paulboone commented on execute_process_mass_data.py:
```diff
@@ -42,6 +42,14 @@
+# sim_data = sys.argv[1]
+# exp_data = sys.argv[2]
+sim_data = 'ALL_MOFS.csv'
+exp_data = 'exp_data_175.csv'
```
we should make this into a simple command line script

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -490,38 +490,25 @@ def save_element_pmf_data(element_pmf_results, stdev, mrange):
     data_frame = pd.DataFrame(element_pmf_results)
     data_frame.to_csv('saved_element_pmfs/%s_stdev_%s_mrange_%s.csv' % (stdev, mrange, timestamp), sep='\t')
```
looks like a bad rebase

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -383,17 +383,31 @@ def bin_compositions(gases, bins, list_of_arrays, all_array_pmf_results):
         dict_temp = {'MOF_Array' : array}
         array_name = '%s' % ' '.join(array)
+
+        # Calculate Aboslute KLD
```
Aboslute

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -467,6 +467,58 @@ def save_element_pmf_data(element_pmf_results, stdev, mrange):
+            num_elements = num_elements
+        else:
+            i = 1
+            num_elements = len(array)
```
repeated again

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -467,6 +467,58 @@ def save_element_pmf_data(element_pmf_results, stdev, mrange):
+        if len(array_name) <= 40:
+            filename = "saved_array_pmfs_unbinned/%s/%s.csv" % (timestamp, array_name)
+        else:
+            filename = "saved_array_pmfs_unbinned/%s/%s.csv" % (timestamp, str(array_id_dict[array_name]))
```
delete; always store ids

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -509,7 +561,10 @@ def save_binned_array_pmf_data(gases, list_of_arrays, bins, binned_probabilities
+            if len(array_name) <= 40:
+                filename = "saved_array_pmfs_binned/%s/%s_%s.csv" % (timestamp, array_name, str(gas))
+            else:
+                filename = "saved_array_pmfs_binned/%s/%s_%s.csv" % (timestamp, str(array_id_dict[array_name]), str(gas))
```
remove this like above

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -564,10 +619,11 @@ def plot_binned_array_pmf_data(gases, list_of_arrays, bins, binned_probabilities
             if len(array) <= 40:
-                plt.savefig("figures/%s/%s_%s.png" % (timestamp, array, str(gas)))
+                plt.title('Array: %s, Gas: %s' % (array, gas))
+                plt.savefig("saved_array_pmfs_binned_figures/%s/%s_%s.png" % (timestamp, array, str(gas)))
```
delete extra condition like above

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -564,10 +619,11 @@ def plot_binned_array_pmf_data(gases, list_of_arrays, bins, binned_probabilities
             else:
-                plt.savefig("figures/%s/%s_%s.png" % (timestamp, 'Array #'+array_id_dict[array]+' (See Key)', str(gas)))
+                plt.title('Array: %s, Gas: %s' % (array_id_dict[array], gas))
+                plt.savefig("saved_array_pmfs_binned_figures/%s/%s_%s.png" % (timestamp, 'Array#'+array_id_dict[array]+'_(See Key)', str(gas)))
```
use second string interpolation instead of '' + '' + ''; remove str() around gas if unnecessary

paulboone commented on execute_process_mass_data.py:
```diff
@@ -98,7 +98,7 @@
 # --------------------------------------------------
 element_pmf_results_df = pd.DataFrame(data=element_pmf_results)
-save_element_pmf_data(element_pmf_results_df, stdev, mrange)
+# save_element_pmf_data(element_pmf_results_df, stdev, mrange)
```
commenting out / uncommenting lines like this is a sign that we need a config file or a command or something?

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -503,20 +503,31 @@ def save_unbinned_array_pmf_data(gases, list_of_arrays, all_array_pmf_results):
     # Generate array data and write to file
+    header = []
+    for gas in gases:
+        header.extend([str(gas)])
```
eliminate str() if not necesssary

header += gas

header = gases

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -503,20 +503,31 @@ def save_unbinned_array_pmf_data(gases, list_of_arrays, all_array_pmf_results):
+    comps_to_save_alt =[]
     for row in all_array_pmf_results:
         comps_to_save.append([{'%s' % gas: row[gas]} for gas in gases])
+        comps_to_save_alt.append(np.transpose([row[gas] for gas in gases]))
```
maybe a comment for what's happening in english? is the np.transpose necessary?

birdday committed:
Changed assignment and saving of array ids into a separate function.

Modified the functions where it was being used (i.e. I removed the lines of code which are identical to the function and passed in the dictionary form of the array ids instead). Note that in these functions I also removed the arbitrary conditions regading whether or not the array name was less than 40 characters. Instead, the array id is not used every time. Necessary changes to execute script also included.

birdday committed:
Updated new array id function. Switched yaml file to correct one. Fixed line continuation.

Added missing function variables in the call in execute. Shortened header assignment in save_unbinned_array_pmf_data.

birdday committed:
Moved the timestamp creation outside the functions for consistency.

birdday committed:
Removed extra comment.

birdday committed:
Fixed spelling errors in comments.

birdday committed:
Modified string interpolation. Changed all instances of 'str(gas)' to 'gas'.

birdday committed:
Changed syntax for adding dictionary keys from x.update({key: val}) to x[key] = val to reduce computational overhead. Note that this only works when adding a single key. calculate_element_pmf still uses the update syntax - to be changed in next commit along with other small changes.

birdday committed:
Changed syntax for adding dictionary keys in calculate_element_pmf function. Also uncommented the PMF_Exact key, although it is not used in the reminder of the code. Tests planned for that later.

birdday committed:
Removed import * and replaced with appropriate functions. Removed bulky comment following import.

birdday committed:
Changed '%s' % ' '.join() string interpolations.

birdday committed:
Removed unnecessary rounding of bin points.

birdday committed:
Removed final return().

birdday committed:
Fixed 'suture' typo.

paulboone commented on mof_array/pmf/process_mass_data.py:
```diff
@@ -1,474 +1,624 @@
+    for array in list_of_arrays:
+        if num_elements == len(array):
+            i += 1
+            num_elements = num_elements
```
delete this

birdday committed:
Removed unnecessary x=x line in assign_array_ids function.

birdday committed:
Changes imports to follow PEP8 formatting. Added comment to comps to save alt, and deleted unnecessary use of np.transpose().
